add DNSSEC validation to the recursive resolver

japaric opened the pull request:
this is an initial implementation of DNSSEC validation that maximally reuses the existing `DnssecDnsHandle` logic. 

it passes the "happy path" conformance tests but there's more work to be done to pass the tests that exercise the failure modes. namely, fixing the bug reported in #2252 . also, handling the CD (Checking Disabled) bit has not yet been implemented

~~this PR depends on #2245 so I have set that PR's branch as the base for this PR.~~

~~this PR depends on PRs #2257 , #2258 and #2268~~ all those PRs have been merged

~~to avoid this being merged into the wrong branch I'm opening this as a *draft PR* but **it is ready for review**~~ this is now ready to be merged

closes #2194 
closes #2234

japaric commented on crates/proto/src/serialize/txt/parse_rdata.rs:
```diff
@@ -88,6 +88,9 @@ impl RDataParser for RData {
             RecordType::TXT => Self::TXT(txt::parse(tokens)?),
             RecordType::SIG => return Err(ParseError::from("parsing SIG doesn't make sense")),
+            #[cfg(feature = "dnssec")]
+            RecordType::DNSKEY => Self::DNSSEC(DNSSECRData::DNSKEY(dnskey::parse(tokens)?)),
```
IIUC, hickory-dns is not meant to work with pre-signed zone files like `nsd` does so when this `Parser` API is used to load a zone file into a "file" authority we should continue to reject DNSKEY records.

IOW, I should probably add some flag here that determines whether DNSKEY records should be accepted or not. this `Parser` is also used to parse a "trust anchor" file which is a newline separated list of DNSKEY records; in that case we want the `Parser` to (only) parse DNSKEY records. when we use `Parser` in service of a file authority we want to reject DNSKEY records like the original code was doing.

djc replied:
In our usage, is there any case where we need to parse both `DNSKEY` records and other kinds of records? If not, we should just split the parser?

japaric replied:
Well, `unbound` supports trust anchor files that contain both `DNSKEY` and `DS` records. Meaning that `unbound` can complete DNSSEC validation with just a DS record in the trust anchor file.

In their current state, `DnssecDnsHandle` and `TrustAnchor` only support DNSSEC validation using public key material (`DNSKEY` information) so a trust anchor file with just DS records would not work and should be rejected.

So, yeah. It's probably best to have separate parsers: the general `Parser` for regular zone files and a `TrustAnchorParser` that only accepts newline separated DNSKEY records (and maybe also DS records in the future).

japaric replied:
I have submitted this whole separate parser idea in #2257

japaric commented on crates/recursor/src/recursor.rs:
```diff
@@ -307,366 +301,106 @@ impl Recursor {
+                let do_bit = true;
+                this.resolve(query, Instant::now(), do_bit)
+                    .map_ok(|lookup| {
+                        // XXX `DnssecDnsHandle` will only look at the answer section of the message so
```
this implementation ended up being a bit "forced" as it assumes that this interface will only be used by the `DnssecDnsHandle` newtype, which is the case.

I feel like changing the `DnsHandle` interface would make *this* code reuse of `DnssecDnsHandle` cleaner but `DnsHandle` is used in other places so that would be a much larger change.

Similarly, I considered creating a new "lookup only" interface alternative to `DnsHandle` that would be used as a trait bound in `DnssecDnsHandle` but then I wasn't sure which of the other existing `DnsHandle` implementations should move to new interface and which others should keep using `DnsHandle`.

japaric commented on crates/server/Cargo.toml:
```diff
@@ -53,7 +53,7 @@ dnssec = ["hickory-recursor?/dnssec"]
 resolver = ["hickory-resolver"]
 sqlite = ["rusqlite"]
-toml = ["dep:basic-toml"]
+toml = ["dep:toml"]
```
as I mentioned in https://github.com/hickory-dns/hickory-dns/issues/2194#issuecomment-2176497799 `basic-toml` cannot deserialize into enum variants that have fields so I switched the toml library to the `toml` crate

djc replied:
IMO that should be a separate commit.

japaric commented on conformance/packages/dns-test/src/docker/hickory.Dockerfile:
```diff
@@ -1,15 +1,18 @@
 # any directory inside the `dns-test` repository
 COPY ./src /usr/src/hickory
-RUN cargo install --path /usr/src/hickory/bin --features recursor,dnssec-ring --debug && \
+RUN cargo install --path /usr/src/hickory/bin --features recursor,dnssec-openssl --debug && \
```
alternatively, we can continue using the `dnssec-ring` feature and instead switch the algorithm used to sign zone files in conformance tests from RSASHA1_NSEC3 to RSASHA256. any preference?

currently the algorithm is [hardcoded in the `dns-test` library](https://github.com/hickory-dns/hickory-dns/blob/02a8668fd54be5c3b1a9f8d8e79393229ffea2be/conformance/packages/dns-test/src/name_server.rs#L221) but we'll probably want to make that configurable on a test basis. that would let us add a regression test for #2252 if we do switch the algorithm to RSASHA256

djc replied:
We hsould probably switch to SHA256 anyway since SHA-1 is deprecated. Would definitely prefer to continue relying on ring (and maybe aws-lc-rs later?) over OpenSSL.

djc commented on crates/recursor/src/recursor.rs:
```diff
@@ -82,7 +59,7 @@ impl RecursorBuilder {
     ///
     /// This will panic if the roots are empty.
-    pub fn build(&self, roots: impl Into<NameServerConfigGroup>) -> Result<Recursor, ResolveError> {
+    pub fn build(&self, roots: impl Into<NameServerConfigGroup>) -> Recursor {
```
Nit: would like to extract this change into a separate commit.

djc commented on crates/recursor/src/common.rs:
```diff
@@ -0,0 +1,149 @@
+use std::time::Instant;
```
Not a big fan of files that are this small. Maybe inline the contents of `common` and `dnssec_policy` into `lib.rs`?

Also please keep moving code around in a separate commit.

djc commented on crates/recursor/src/recursor.rs:
```diff
@@ -47,10 +55,10 @@ impl RecursorBuilder {
     #[cfg(feature = "dnssec")]
-    pub fn security_aware(&mut self, security_aware: bool) -> &mut Self {
-        self.security_aware = security_aware;
+    pub fn dnssec_policy(&mut self, dnssec_policy: DnssecPolicy) -> &mut Self {
```
Would also like a separate commit for the introduction of `DnssecPolicy`.

djc commented on crates/recursor/src/recursor.rs:
```diff
@@ -59,17 +67,12 @@ impl RecursorBuilder {
-        #[cfg(feature = "dnssec")]
-        let security_aware = self.security_aware;
-
+    pub fn build(&self, roots: impl Into<NameServerConfigGroup>) -> Result<Recursor, Error> {
```
If we're going to flip-flop on the fallibility of this method, let's just keep it fallible?

djc commented on crates/recursor/src/recursor.rs:
```diff
@@ -266,8 +301,106 @@ impl Recursor {
+}
+
+enum RecursorEither {
+    NonValidating(RecursorDnsHandle),
```
Why doesn't the non-validating variant need a cache?

I dislike `Either` naming because there is no semantic meaning. Can you come up with a better name for this?

djc commented on crates/recursor/src/recursor_dns_handle.rs:
```diff
@@ -138,31 +138,7 @@ impl RecursorDnsHandle {
-
-                lookup.ok_or_else(|| Error::from("no records found"))
-            }
+            Ok(r) => common::insert_response(r, Some(ns.zone()), &self.record_cache, query, now),
```
Please make this a separate commit, too.

djc commented on crates/recursor/src/recursor_dns_handle.rs:
```diff
@@ -27,9 +26,10 @@ use crate::{
 /// Set of nameservers by the zone name
 type NameServerCache<P> = LruCache<Name, RecursorPool<P>>;
 
+#[derive(Clone)]
```
Why is this necessary? Are the `RecursorPool` and `DnsLru` cheap to clone? Why not wrap the entire `RecursorDnsHandle` in an `Arc`?

japaric replied:
> Why is this necessary?

`DnsHandle` is a subtrait of `Clone` and we need to implement `DnsHandle` to put this type inside `DnssecDnsHandle<_>`

> Are the RecursorPool and DnsLru cheap to clone?

`RecursorPool` has several `Arc` fields and  `DnsLru` is esentially an `Arc<Mutex<_>>` so, yes

> Why not wrap the entire RecursorDnsHandle in an Arc?

doable but that would result in a nested `Arc` which ... I guess it's fine? I don't think either way has much impact on performance, either CPU or memory usage.

djc replied:
I guess it's fine, no need to change it on my account.

djc commented on crates/proto/src/serialize/txt/rdata_parsers/dnskey.rs:
```diff
@@ -0,0 +1,96 @@
+
+#[allow(deprecated)]
+pub(crate) fn parse<'i, I: Iterator<Item = &'i str>>(mut tokens: I) -> ParseResult<DNSKEY> {
+    let flags_str: &str = tokens
```
Nit: let's skip the unnecessary type annotations (in new code)?

djc commented on crates/proto/src/serialize/txt/rdata_parsers/dnskey.rs:
```diff
@@ -0,0 +1,96 @@
+        .next()
+        .ok_or_else(|| ParseError::from(ParseErrorKind::Message("algorithm not present")))?;
+
+    let flags: u16 = flags_str.parse()?;
```
I'd prefer calling this as `u16::from_str(flags_str)` (same for other uses of `.parse()`).

djc commented on crates/server/src/store/recursor/config.rs:
```diff
@@ -78,3 +80,122 @@ fn ns_cache_size_default() -> usize {
+    },
+}
+
+impl Default for DnssecPolicyConfig {
```
Can this be derived?

djc commented on crates/server/src/store/recursor/config.rs:
```diff
@@ -78,3 +80,122 @@ fn ns_cache_size_default() -> usize {
+
+impl DnssecPolicyConfig {
+    pub(crate) fn load(&self) -> Result<DnssecPolicy, String> {
+        let policy = match self {
```
Nit: inline the `policy` binding?

djc commented on crates/server/src/store/recursor/config.rs:
```diff
@@ -78,3 +80,122 @@ fn ns_cache_size_default() -> usize {
+
+    let mut trust_anchor = TrustAnchor::new();
+    for (key, rrset) in records {
+        if key.record_type == RecordType::DNSKEY {
```
Nit: invert and continue early?

djc commented on crates/server/src/store/recursor/authority.rs:
```diff
@@ -172,4 +172,12 @@ impl LookupObject for RecursiveLookup {
     }
+
+    fn dnssec_validated(&self) -> bool {
+        // XXX what should be the behavior if the lookup yields no records?
```
Does this need discussion?

japaric replied:
I think it needs a better understanding of the RFCs and an accompanying conformance test.

I think the condition mentioned here would happen if a query yields an empty answer section (NXDOMAIN), DNSSEC validation is enabled but both NSEC and NSEC3 are disabled. "does the AD flag get set in the response or not?" is what needs to be researched, either in RFCs or by testing what other implementations do.

I have that in my TODO list. I can also open issues for missing / pending conformance tests if you think that's worthwhile.

djc replied:
If it's on your list I'm fine merging it as is. Just wanted to mark it with a discussion to get a bit more context on how/when this will be followed up on.

japaric commented:
Addressed all review comments, I believe. Note that this version includes two other open PRs as cherry picks. I'll remove them once those PRs have been merged

djc commented on crates/recursor/src/recursor_dns_handle.rs:
```diff
@@ -0,0 +1,295 @@
+use std::net::SocketAddr;
```
This doesn't look like it belongs in this commit.

japaric replied:
good catch! fixed

japaric commented on crates/recursor/src/lib.rs:
```diff
@@ -54,6 +56,15 @@ pub enum DnssecPolicy {
+    #[cfg(feature = "dnssec")]
+    ValidateWithStaticKey {
+        /// set to `None` to use built-in trust anchor
+        trust_anchor: Option<TrustAnchor>,
```
I like that idea.

It does require changes elsewhere because this `trust_anchor` is used to initialize a `DnssecDnsHandle` and its constructor expects a `TrustAnchor` by value ... which then immediately wraps in an `Arc` for storage in one of its fields.

I submitted the `Arc` changes in a separate commit

japaric commented:
rebased to fix merge conflict

note that I have cherry picked PR #2268 because `dns_test::resolver::terminate_hickory_works` started failing after the configuration file changed ("expose DnssecPolicy in RecursorConfig")

until the 3 cherry picked PRs have been merged, I'll keep this PR in the draft state but it is ready for review

japaric commented:
amended a commit ("expose DnssecPolicy in RecursorConfig") to fix a clippy warning

yonas commented:
@japaric I just tested the recurser. It doesn't pickup the `--roots` CLI flag.

japaric commented:
@yonas thanks for testing. that sounds like it may be a pre-existing bug.

the end-to-end tests here all configure the root hints in `named.toml`; see the `conformance/packages/dns-test/src/templates/hickory.resolver.toml.jinja` file in this branch / PR. 

did you set up a `name.toml` configuration file and additionally passed the `--roots` flag to `hickory-dns`? I'm not sure what the precedence rules are in that particular case since I didn't touch CLI code in this PR. if that sort of usage does not produce a warning / error then could you open an issue about that? because to me that sounds like it should.

relatedly, we could use some docs on setting up `hickory-dns` as a recursive resolver. I floated the idea in https://github.com/hickory-dns/hickory-dns.github.io/issues/2 but I'm sure how well publicized the book currently is. maybe those docs can go in the README for now, if that has more visibility. in any case, it would be good to fix #2252 before the DNSSEC feature gets more publicity, IMO.

wait, `hickory-dns` does not have a `--roots` flags. are you using `util/src/bin/recurse.rs`?  that tool has not been touched in this PR. I don't know what's the plan with `hickory-util` so DNSSEC has not been exposed there.

yonas commented:
No, I did not setup `name.toml`

Yes, I used the `recurse` binary.

Which binary does this PR affect?

japaric commented:
> Which binary does this PR affect?

`hickory-dns` (DNS server) which resides in `bin/src/hickory-dns.rs` and you can build with `cargo build -p hickory-dns`. the artifact will then be in `target/debug/hickory-dns`. you probably want to run that with `-p 1053 -c my-named.toml` because the defaults require root permissions (to bind port 53 and to create a file in `/etc/named.toml`). if you need help setting that up, I would suggest asking in the [discord server](https://discord.gg/89nxE4n) to keep this thread on topic.

bluejekyll commented:
@japaric is this ready to merge in your opinion? I know there was a lot of commentary back and forth.

justahero commented on conformance/packages/dns-test/src/templates/hickory.resolver.toml.jinja:
```diff
@@ -1,5 +1,4 @@
 zone = "."
 zone_type = "Hint"
-stores = { type = "recursor", roots = "/etc/root.hints", security_aware = true }
-enable_dnssec = {{ use_dnssec }}
```
If my understanding is correct the `enable_dnssec` is not used by the `Resolver`, the DNSSEC logic is handledby the given `DnssecPolicy` here? Does this in any way affect the example`toml` files in `tests/test-data/test_configs` &  should they be updated as well?

japaric replied:
`enable_dnssec` is only used by the `file` and `sqlite` type stores (which must be used with the `Primary` and `Secondary` `zone_type`s). it is not used by the `recursor` (`Hint`) nor the `forward`-type stores. 

IOW, this `enable_dnssec` had no effect in this file. I added a while ago when I did not understand the structure of the `named.toml` file -- which then prompted me to suggest RFC #2195

So, you should not understand `enable_dnssec = true` must now be replaced with `dnssec_policy = Something` but rather as the `enable_dnssec`, which was a no-op, is getting removed (clean-up); *and* the use of `security_aware` is being replaced with `dnssec_policy`.

Therefore, the `toml` files under `test_configs` do not need to be updated as part of this PR.

japaric commented:
rebased to fix a merge conflict caused by a docstring change

@bluejekyll yes, this is ready to be merged :rocket:

djc commented:
This failed CI on main:

https://github.com/hickory-dns/hickory-dns/actions/runs/9777369448/job/26991905906#step:5:2056

japaric commented:
that's a known spurious failure that happens very infrequently: https://github.com/ferrous-systems/dnssec-tests/issues/55

I'll port that issue over here
